### ISSUE TITLE
terraform crash when using an aliased provider and data block 

### ISSUE THREAD

**Issue by joshschmitter** [hashicorp/terraform#25311]
### Terraform Version

```
0.13.0 beta2 
```

### Terraform Configuration Files

```hcl
provider "azurerm" {
  subscription_id = "zzzzz-zzzzz-zzzzzz-zzzzzz" #redacted
  version         = "~> 2.1.0"
  features {}
}

provider "azurerm" {
  alias           = "platform_kv_subscription" 
  subscription_id = "xxxx-xxxxx-xxxxxx-xxxxxx" #redacted
  version         = "~> 2.1.0"
  features {}
}

data "azurerm_key_vault" "platform_shared" {
  provider                        = "platform_kv_subscription"
  name                             = "myplatformkv"
  resource_group_name = "myresourcegroup"
}
```

### Debug Output
https://gist.github.com/joshschmitter/b881082e6286f1be5adf8fdb7e30c1bf#file-debug_output-log

### Crash Output
https://gist.github.com/joshschmitter/b881082e6286f1be5adf8fdb7e30c1bf#file-crash-log

### Expected Behavior
~~terraform should have downloaded providers and data-block data.~~

*Update*: since the provider reference in the data block was missing the provider type (`azurerm`), terraform should have thrown an error: `Error: no provider exists with the given name`

### Actual Behavior
terraform crashed.

```
panic: must contain only letters, digits, and dashes, and may not use leading or trailing dashes
```

### Steps to Reproduce
1. `terraform init`
2. `terraform apply` 

### Additional Context
The reference to the provider was incorrect, missing the provider type of `azurerm` before the provider alias name.

**Comment by joshschmitter** [comment]
Looks like the underlying issue was that, in the data block, the provider value should have been `azurerm.platform_kv_subscription`.  I was missing `azurerm.`

This was revealed by rolling back to terraform 0.12.26 and running the same code.  The result was:

```
Initializing modules...

Initializing the backend...

Initializing provider plugins...
- Checking for available provider plugins...

Provider "platform_kv_subscription" not available for installation.

A provider named "platform_kv_subscription" could not be found in the Terraform Registry.

This may result from mistyping the provider name, or the given provider may
be a third-party provider that cannot be installed automatically.

In the latter case, the plugin must be installed manually by locating and
downloading a suitable distribution package and placing the plugin's executable
file in the following directory:
    terraform.d/plugins/darwin_amd64

Terraform detects necessary plugins by inspecting the configuration and state.
To view the provider versions requested by each module, run
"terraform providers".


Error: no provider exists with the given name

```

**Comment by alisdair** [comment]
@joshschmitter Thanks for reporting this with a clear reproduction case!

There's a fix for this crash in #25408. The output now correctly displays the following diagnostics:

```shellsession
$ terraform init
There are some problems with the configuration, described below.

Terraform found syntax errors in the configuration that prevented full
initialization. If you've recently upgraded to Terraform v0.13 from Terraform
v0.11, this may be because your configuration uses syntax constructs that are no
longer valid, and so must be updated before full initialization is possible.

Manually update your configuration syntax, or install Terraform v0.12 and run
terraform init for this configuration at a shell prompt for more information
on how to update it for Terraform v0.12+ compatibility.

Warning: Quoted references are deprecated

  on main.tf line 15, in data "azurerm_key_vault" "platform_shared":
  15:   provider                        = "platform_kv_subscription"

In this context, references are expected literally rather than in quotes.
Terraform 0.11 and earlier required quotes, but quoted references are now
deprecated and will be removed in a future version of Terraform. Remove the
quotes surrounding this reference to silence this warning.


Error: Invalid provider local name

  on main.tf line 15, in data "azurerm_key_vault" "platform_shared":
  15:   provider                        = "platform_kv_subscription"

platform_kv_subscription is an invalid provider local name: must contain only
letters, digits, and dashes, and may not use leading or trailing dashes
```

As the warning diagnostic recommends, the correct syntax for the `provider` attribute is without quotes:

```hcl
data "azurerm_key_vault" "platform_shared" {
  provider = azurerm.platform_kv_subscription
}
```
